### PR TITLE
[#35] - Pass configuration to PathManager

### LINKED ISSUE #35: Pass configuration to PathManager
# Scenario Summary
Refactor configuration loading of Managers i.e. `PathManager` and `IndexManager`, because at the moment configuration loading happens wrongly.
Update config properties file hierarchy in a 1:1 fashion between manager and config node. 

# Problem Statement
Many managers call other managers functions inside but at the moment, those are implemented like Object. 
The result is that every manager read its own configuration, but it is made in the wrong way.
We want to refactor manager into case classes and build a config file reflecting the dependencies between a manager.

# Proposed Solution

Refactor the object `PathManager` and `IndexManager` adding a `Config` as a parameter.
Update config file properties hierarchy.

### PR DESCRIPTION
**Pull request by giorgio84**
# New features and improvements

- `PathManager`, `IndeManager` and `OSMAdministrativeBoundariesLoader` turned into case classes
- Removed previous side effects contained in Managers
- Improved config file hierarchy. Now it reflects dependencies between managers   

# Breaking changes

No

# Bug fixes

Wrong configuration loading from `IndexManager`, `PathManager` and `OSMAdministrativeBoundariesLoader`

# Related issue

Closes #35

### CODE REVIEW

File: src/test/scala/it/agilelab/bigdata/gis/domain/loader/OSMManagerSpec.scala [hunk 1]
```diff
@@ -2,13 +2,15 @@ package it.agilelab.bigdata.gis.domain.loader
...
 import it.agilelab.bigdata.gis.domain.graphhopper.GPSPoint
+import it.agilelab.bigdata.gis.domain.managers.OSMManager
 import it.agilelab.bigdata.gis.domain.models.ReverseGeocodingResponse
 import org.scalatest.{BeforeAndAfterAll, FlatSpec, Matchers}
 
 class OSMManagerSpec extends FlatSpec with Matchers with BeforeAndAfterAll {
 
-  val conf: Config = ConfigFactory.load()
-  val osmManager: OSMManager = OSMManager(conf)
+  val conf: Config = ConfigFactory.load("reference.conf")
```
Comment (andr3a87) [inline src/test/scala/it/agilelab/bigdata/gis/domain/loader/OSMManagerSpec.scala:11]: is it necessary specify `reference.conf`?

File: src/main/scala/it/agilelab/bigdata/gis/domain/managers/IndexManager.scala [hunk 2]
```diff
@@ -1,18 +1,23 @@
...
-import it.agilelab.bigdata.gis.domain.managers.ManagerUtils.BoundariesPos.{CITIES, COUNTIES, COUNTRY, REGIONS}
 
+import java.io.File
 import scala.collection.immutable
 
+case class IndexManager(conf: Config) extends Configuration with Logger {
 
-object IndexManager extends Logger {
+  val indexConfig: IndexManagerConfiguration = IndexManagerConfiguration(conf)
+  val pathManager : PathManager = PathManager(conf.getConfig(ConfigurationProperties.PATH.value))
```
Comment (andr3a87) [inline src/main/scala/it/agilelab/bigdata/gis/domain/managers/IndexManager.scala:18]: can you use a ConfigurationReader?

File: src/main/scala/it/agilelab/bigdata/gis/domain/managers/IndexManager.scala [hunk 3]
```diff
@@ -1,18 +1,23 @@
...
+  val boundariesLoader : OSMAdministrativeBoundariesLoader = OSMAdministrativeBoundariesLoader(conf.getConfig(ConfigurationProperties.BOUNDARY.value), pathManager)
```
Comment (andr3a87) [inline src/main/scala/it/agilelab/bigdata/gis/domain/managers/IndexManager.scala:19]: can you use a ConfigurationReader?

File: src/test/scala/it/agilelab/bigdata/gis/domain/loader/OSMManagerSpec.scala (hunk 1 above)
Comment (giorgio84) [reply src/test/scala/it/agilelab/bigdata/gis/domain/loader/OSMManagerSpec.scala:11]: Actually no, it was put it just for clarity.

File: src/main/scala/it/agilelab/bigdata/gis/domain/managers/IndexManager.scala (hunk 2 above)
Comment (giorgio84) [reply src/main/scala/it/agilelab/bigdata/gis/domain/managers/IndexManager.scala:18]: Right, I will go to put `pathConf` as parameter calculated from `IndexManagerConfiguration`

File: src/main/scala/it/agilelab/bigdata/gis/domain/managers/IndexManager.scala (hunk 3 above)
Comment (giorgio84) [reply src/main/scala/it/agilelab/bigdata/gis/domain/managers/IndexManager.scala:19]: same as above. I will go to fix also `OSMManager`, that at the moment is doing the same thing

**Review by andr3a87 — APPROVED** [review]

**Pull request merged**